lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,625 @@
+        | { status: "CREDENTIAL_MISSING_ERROR" }
+        | { status: "GENERATED_OPTIONS_ID_MISSING_ERROR" }
+        | { status: "RECOVER_ACCOUNT_TOKEN_MISSING_ERROR" }
+        | { status: "RECOVER_ACCOUNT_TOKEN_INVALID_ERROR" }
+    >;
+
+    // used internally for creating a passkey during password reset flow or when adding a credential to an existing user
+    // email will be taken from the options
+    // no need for recoverAccountToken, as that will be used upstream
+    // (in consumeRecoverAccountToken invalidating the token and in registerOptions for storing the email in the generated options)
+    registerPasskey(input: {
+        webauthnGeneratedOptionsId: string;
+        credential: {
+            id: string;
+            rawId: string;
+            response: {
+                clientDataJSON: string;
+                attestationObject: string;
+                transports?: ("ble" | "cable" | "hybrid" | "internal" | "nfc" | "smart-card" | "usb")[];
+                userHandle: string;
+            };
+            authenticatorAttachment: "platform" | "cross-platform";
+            clientExtensionResults: Record<string, unknown>;
+            type: "public-key";
+        };
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | {
+              status: "OK";
+              user: User;
+              recipeUserId: RecipeUserId;
+          }
+        | { status: "CREDENTIAL_MISSING_ERROR" }
+        | { status: "WRONG_CREDENTIALS_ERROR" }
+        | { status: "GENERATED_OPTIONS_ID_MISSING_ERROR" }
+    >;
+
+    // this function is meant only for creating the recipe in the core and nothing else.
+    // we added this even though signUp exists cause devs may override signup expecting it
+    // to be called just during sign up. But we also need a version of signing up which can be
+    // called during operations like creating a user during password reset flow.
+    createNewRecipeUser(input: {
+        webauthnGeneratedOptionsId: string;
+        credential: {
+            id: string;
+            rawId: string;
+            response: {
+                clientDataJSON: string;
+                attestationObject: string;
+                transports?: ("ble" | "cable" | "hybrid" | "internal" | "nfc" | "smart-card" | "usb")[];
+                userHandle: string;
+            };
+            authenticatorAttachment: "platform" | "cross-platform";
+            clientExtensionResults: Record<string, unknown>;
+            type: "public-key";
+        };
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | {
+              status: "OK";
+              user: User;
+              recipeUserId: RecipeUserId;
+          }
+        | { status: "EMAIL_ALREADY_EXISTS_ERROR" }
+        | { status: "WRONG_CREDENTIALS_ERROR" }
+        | { status: "CREDENTIAL_MISSING_ERROR" }
+        | { status: "GENERATED_OPTIONS_ID_MISSING_ERROR" }
+    >;
+
+    verifyCredentials(input: {
```

Comment:
this function has identical func signature as registerPasskeys. Are both, this and the other function needed?

Reply:
yes. one logs in the user, the other adds a credential to the database

Reply:
Thats too high level of an explanation.. what will be the exact diff between the implementation of them? Its not like the signIn will create a session - so it doesn't "log in" the user.

Reply:
They do different things. It's like sign up vs sign in in `emailpassword`. They both take a password and an email but do different things. 

I replicated what was on the emailpassword recipe: 
https://github.com/supertokens/supertokens-node/blob/ced57b11260144d1e1b89039db3c1bdbbc258318/lib/ts/recipe/emailpassword/types.ts#L148

registerCredentials (registerPasskeys) will be used on sign up and for adding a passkey to a logged in user.

Am i missing something ?

Reply:
So verifyCredentials can only be used by options created from signInOptions, whereas registerPasskeys can only be used by options created from registerOptions?

Reply:
yes. but maybe let's chat.